File: CommonFiles/Common_Files_VB/Wrapper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Immutable
Imports System.Linq
Imports System.Math
Imports System.Runtime.CompilerServices
Imports System.Text

Namespace MetalUp.Express

    Module Program
        Public Sub Main()
            Console.OutputEncoding = Encoding.UTF8
            Console.WriteLine(Display("<Expression>"))
        End Sub

        ' StudentCode - this comment is needed error line number offset
        '<StudentCode>

        '<HiddenCode>

    End Module

    '<Tests>

    Public Module Helpers

        Public Function Display(ByVal obj As Object) As String
            If obj Is Nothing Then
                Return ""
            End If
            If TypeOf obj Is String Then
                Return $"""{obj}"""
            End If
            If TypeOf obj Is Boolean Then
                Return If(DirectCast(obj, Boolean), "true", "false")
            End If
            If TypeOf obj Is ITuple Then
                Dim tuple = DirectCast(obj, ITuple)
                Return Enumerable.Range(0, tuple.Length).Aggregate("(", Function(s, i) s & Display(tuple(i)) & (If(i < (tuple.Length - 1), ",", ")")))
            End If
            Dim type = obj.GetType()
            If type.IsGenericType AndAlso type.GetGenericTypeDefinition() Is GetType(List(Of )) Then
                Dim toDisplay = DirectCast(obj, IEnumerable).Cast(Of Object)().Select(Function(o) Display(o)).ToList()
                Return vbCrLf + $"{{{String.Join(", ", toDisplay)}}}"
            End If
            If TypeOf obj Is IEnumerable Then
                Return "Result is an IEnumerable. Convert to List to display contents"
            End If
            Return obj.ToString()
        End Function

        Public Function ArgString(ParamArray arguments As Object()) As String
            Return arguments.Aggregate("", Function(s, a) s & Display(a) & ", ").TrimEnd(" "c, ","c)
        End Function

        Public Function FailMessage(functionName As String, ByVal expected As Object, ByVal actual As Object, ParamArray args As Object()) As String
            Return $"xxxTest failed calling {functionName}({ArgString(args)}) Expected: {Display(expected)} Actual: {Display(actual)}xxx"
        End Function

        Public Function EqualIfRounded(expected As Double, actual As Double) As Boolean
            Dim places = expected.ToString().Length - expected.ToString().IndexOf(".") - 1
            Dim rounded = Math.Round(actual, places)
            Return expected = rounded
        End Function

        <Extension>
        Public Function SetItem(Of T)(ByVal list As List(Of T), ByVal i As Integer, ByVal value As T) As List(Of T)
            Return list.Take(i).Append(value).Concat(list.Skip(i + 1)).ToList()
        End Function

        <Extension>
        Public Function InsertItem(Of T)(ByVal list As List(Of T), ByVal i As Integer, ByVal value As T) As List(Of T)
            Return list.Take(i).Append(value).Concat(list.Skip(i)).ToList()
        End Function

        <Extension>
        Public Function RemoveItem(Of T)(ByVal list As List(Of T), ByVal i As Integer) As List(Of T)
            Return list.Take(i).Concat(list.Skip(i)).ToList()
        End Function

    End Module

End Namespace

File: CommonFiles/Common_Files_CS/Wrapper.cs
using static MetalUp.Express.Helpers;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using static System.Math;
using System.Runtime.CompilerServices;
using System.Text;

namespace MetalUp.Express
{
    public class Wrapper
    {
        static void Main()
        {
            Console.OutputEncoding = Encoding.UTF8;
            Console.WriteLine(Display("<Expression>"));
        }
//COMMENTS BELOW MUST BE OUTDENTED TO MAX

// StudentCode - this comment is needed error line number offset
//<StudentCode>
//<HiddenCode>
//<Tests>
    }

    public static class Helpers
    {
        public static string Display(object obj)
        {
            if (obj is null) return "";
            if (obj is string) return $@"""{obj}""";
            if (obj is Boolean) return (Boolean)obj ? "true" : "false";
            if (obj is ITuple)
            {
                var tuple = (ITuple)obj;
                return Enumerable.Range(0, tuple.Length)
                    .Aggregate("(", (s, i) => s + Display(tuple[i]) + (i < (tuple.Length - 1) ? ", " : ")"));
            }
            var type = obj.GetType();
            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(List<>))
            {
                var toDisplay = ((IEnumerable)obj).Cast<object>().Select(o => Display(o)).ToList();
                return $"\n{{{string.Join(", ", toDisplay)}}}";
            }
            if (obj is IEnumerable)
            {
                return "Result is an IEnumerable. Convert to List to display contents";
            }
            return obj.ToString();
        }

        public static string ArgString(params object[] arguments) => arguments.Aggregate("", (s, a) => s + Display(a) + ", ").TrimEnd(' ', ',');

        public static string FailMessage(string functionName, object expected, object actual, params object[] args) =>
             $"xxxTest failed calling {functionName}({ArgString(args)}) Expected: {Display(expected)} Actual: {Display(actual)}xxx";

        public static bool EqualIfRounded(double expected, double actual) =>
            expected == Round(actual, expected.ToString().Length - expected.ToString().IndexOf(".") - 1);
        public static List<T> SetItem<T>(this List<T> list, int i, T value) =>
                list.Take(i).Append(value).Concat(list.Skip(i + 1)).ToList();

            public static List<T> InsertItem<T>(this List<T> list, int i, T value) =>
                list.Take(i).Append(value).Concat(list.Skip(i)).ToList();

            public static List<T> RemoveItem<T>(this List<T> list, int i) =>
                list.Take(i).Concat(list.Skip(i+1)).ToList();
    }
}